src/main/com/fulcrologic/fulcro/algorithms/normalized_state_helpers.cljc

```diff
@@ -0,0 +1,502 @@
+
+   The optional `cascade` parameter is a set of keywords that represent edges that should cause recursive deletes
+   (i.e. it indicates edge names that *own* something, indicating it is safe to remove those entities as well).
+
+   Returns the new state map with the entity(ies) removed."
+
+  ([state-map ident]
+   [map? eql/ident? => map?]
+   (remove-entity* state-map ident #{}))
+
+  ;;TODO implement the cascading feature
+
+  ([state-map ident cascade]
+   [map? eql/ident? (s/coll-of keyword? :kind set?) => map?]
+
+   (reduce #(prune-ident %1 %2 ident)
+           (state-after-top-level-ident-dissoc state-map ident)
+           (entity-path-value-map-after-top-level-dissoc state-map ident))))
+
+
+(comment
+
+  (def state {:fastest-car  [:car/id 1]
+              :grandparents [[:person/id 1] [:person/id 2]]
+              :denorm       {:level-1 {:level-2 {:a [[:person/id 1] [:person/id 2]] :b [:person/id 1]}}}
+              :person/id    {1 {:person/name     "person-1"
+                                :person/spouse   [:person/id 2]
+                                :person/email    [:email/id 1]
+                                :person/cars     [[:car/id 1]]
+                                :person/children [[:person/id 3]
+                                                  [:person/id 4]
+                                                  [:person/id 5]]}
+                             2 {:person/name     "person-2"
+                                :person/spouse   [:person/id 1]
+                                :person/cars     [[:car/id 1]
+                                                  [:car/id 2]]
+                                :person/children [[:person/id 3]
+                                                  [:person/id 4]
+                                                  [:person/id 5]]}
+                             3 {:person/name "person-3"}
+                             4 {:person/name     "person-4"
+                                :person/spouse   [:person/id 6]
+                                :person/children [:person/id 7]}
+                             5 {:person/name "person-5"}
+                             6 {:person/id       6
+                                :person/name     "person-6"
+                                :person/spouse   [:person/id 4]
+                                :person/children [:person/id 7]}
+                             7 {:person/name "person-7"}}
+              :car/id       {1 {:car/model  "model-1"
+                                :car/engine [:engine/id 1]}
+                             2 {:car/model  "model-2"
+                                :car/engine [:engine/id 2]}}
+              :engine/id    {1 {:engine/name "engine-1"}
+                             2 {:engine/name "engine-2"}}
+              :email/id     {1 {:email/provider "Google"}
+                             2 {:email/provider "Microsoft"}}})
```

Comment:
Basically, this gets converted into a path-value map like below. 

I need a bit of a help regarding a recursive function which would take this map as an input and return a vector of all idents which are affected via `cascading`. Rest of the implementation details are already there.


```clojure
{[:accounts/id 1 :account/id]    1,
 [:car/id 1 :car/model]          "car-1",
 [:accounts/id 7 :account/id]    7,
 [:person/id 6 :person/cars]     [[:car/id 1] [:car/id 2]],
 [:provider/id 1 :provider/id]   1,
 [:provider/id 4 :provider/name] "Yahoo",
 [:person/id 1 :person/accounts] [[:account/id 1] [:account/id 2]],
 [:person/id 8 :person/spouse]   nil,
 [:person/id 1 :person/email]    [:email/id 1],
 [:accounts/id 4 :account/name]  "account-4",
 [:email/id 5 :email/url]        "5@test.com",
 [:email/id 3 :email/url]        "3@test.com",
 [:car/id 2 :car/engine]         [:engine/id 2],
 [:car/id 1 :car/engine]         [:engine/id 1],
 [:provider/id 4 :provider/id]   4,
 [:car/id 1 :car/id]             1,
 [:accounts/id 3 :account/name]  "account-3",
 [:provider/id 2 :provider/id]   2,
 [:car/id 2 :car/model]          "car-2",
 [:car/id 2 :car/id]             2,
 [:person/id 3 :person/id]       3,
 [:email/id 7 :email/url]        "7@test.com",
 [:provider/id 1 :provider/name] "Google",
 [:person/id 5 :person/accounts] [[:account/id 5]],
 [:email/id 4 :email/id]         4,
 [:person/id 8 :person/cars]     nil,
 [:email/id 6 :email/id]         6,
 [:email/id 7 :email/provider]   [:provider/id 2],
 [:provider/id 3 :provider/id]   3,
 [:engine/id 3 :engine/id]       3,
 [:accounts/id 4 :account/id]    4,
 [:car/id 4 :car/model]          "car-4",
 [:email/id 3 :email/provider]   [:provider/id 3],
 [:person/id 6 :person/email]    [:email/id 6],
 [:person/id 3 :person/cars]     nil,
 [:person/id 4 :person/spouse]   [:person/id 6],
 [:person/id 1 :person/spouse]   [:person/id 2],
 [:person/id 4 :person/email]    [:email/id 4],
 [:car/id 3 :car/model]          "car-3",
 [:person/id 7 :person/spouse]   nil,
 [:person/id 3 :person/spouse]   nil,
 [:email/id 2 :email/url]        "2@test.com",
 [:person/id 8 :person/name]     nil,
 [:engine/id 1 :engine/id]       1,
 [:person/id 4 :person/children] [:person/id 8],
 [:person/id 5 :person/cars]     [[:car/id 4]],
 [:person/id 6 :person/accounts] [[:account/id 6]],
 [:person/id 5 :person/name]     "person-5",
 [:accounts/id 1 :account/name]  "account-1",
 [:email/id 5 :email/provider]   [:provider/id 2],
 [:email/id 8 :email/id]         8,
 [:email/id 2 :email/provider]   [:provider/id 1],
 [:accounts/id 3 :account/id]    3,
 [:person/id 8 :person/accounts] nil,
 [:person/id 4 :person/id]       4,
 [:person/id 5 :person/children] nil,
 [:email/id 1 :email/url]        "1@test.com",
 [:person/id 6 :person/children] [:person/id 7],
 [:person/id 3 :person/email]    [:email/id 3],
 [:email/id 3 :email/id]         3,
 [:person/id 3 :person/children] nil,
 [:person/id 8 :person/children] nil,
 [:person/id 7 :person/children] nil,
 [:car/id 4 :car/engine]         [:engine/id 2],
 [:person/id 7 :person/id]       7,
 [:person/id 1 :person/children]
                                 [[:person/id 3] [:person/id 4] [:person/id 5]],
 [:email/id 7 :email/id]         7,
 [:email/id 4 :email/provider]   [:provider/id 1],
 [:person/id 1 :person/id]       1,
 [:denorm :level-1 :level-2 :b]  nil,
 [:person/id 4 :person/name]     "person-4",
 [:email/id 8 :email/url]        "8@test.com",
 [:person/id 4 :person/accounts] [[:account/id 4]],
 [:accounts/id 2 :account/name]  "account-2",
 [:person/id 6 :person/name]     "person-6",
 [:accounts/id 5 :account/name]  "account-5",
 [:email/id 6 :email/url]        "6@test.com",
 [:car/id 3 :car/engine]         [:engine/id 3],
 [:car/id 3 :car/id]             3,
 [:email/id 4 :email/url]        "4@test.com",
 [:accounts/id 2 :account/id]    2,
 [:person/id 2 :person/children]
                                 [[:person/id 3] [:person/id 4] [:person/id 5]],
 [:person/id 3 :person/name]     "person-3",
 [:fastest-car]                  [:car/id 3],
 [:email/id 1 :email/id]         1,
 [:car/id 4 :car/id]             4,
 [:engine/id 2 :engine/id]       2,
 [:person/id 1 :person/cars]     [[:car/id 1]],
 [:person/id 7 :person/email]    [:email/id 7],
 [:car/id 5 :car/id]             5,
 [:person/id 3 :person/accounts] nil,
 [:person/id 5 :person/spouse]   nil,
 [:engine/id 4 :engine/id]       4,
 [:person/id 6 :person/spouse]   [:person/id 4],
 [:person/id 2 :person/accounts] [[:account/id 3]],
 [:person/id 8 :person/email]    [:email/id 8],
 [:person/id 2 :person/cars]     [[:car/id 2]],
 [:person/id 6 :person/id]       6,
 [:provider/id 3 :provider/name] "Tencent",
 [:accounts/id 6 :account/id]    6,
 [:car/id 5 :car/model]          "car-5",
 [:person/id 2 :person/spouse]   [:person/id 1],
 [:email/id 2 :email/id]         2,
 [:engine/id 1 :engine/model]    "engine-1",
 [:person/id 4 :person/cars]     [[:car/id 4]],
 [:person/id 2 :person/name]     "person-2",
 [:email/id 8 :email/provider]   [:provider/id 3],
 [:email/id 1 :email/provider]   [:provider/id 1],
 [:deceased]                     [[:person/id 8]],
 [:car/id 5 :car/engine]         [:engine/id 4],
 [:denorm :level-1 :level-2 :a]  [[:b 1]],
 [:engine/id 2 :engine/model]    "engine-2",
 [:person/id 7 :person/name]     "person-7",
 [:accounts/id 6 :account/name]  "account-6",
 [:person/id 2 :person/id]       2,
 [:person/id 2 :person/email]    [:email/id 2],
 [:accounts/id 7 :account/name]  "account-7",
 [:person/id 5 :person/id]       5,
 [:person/id 1 :person/name]     "person-1",
 [:accounts/id 5 :account/id]    5,
 [:person/id 8 :person/id]       8,
 [:engine/id 4 :engine/model]    "engine-4",
 [:provider/id 2 :provider/name] "Microsoft",
 [:engine/id 3 :engine/model]    "engine-3",
 [:email/id 5 :email/id]         5,
 [:grandparents]                 [[:person/id 1] [:person/id 2]],
 [:person/id 7 :person/accounts] [[:account/id 7]],
 [:email/id 6 :email/provider]   [:provider/id 3],
 [:person/id 5 :person/email]    [:email/id 5],
 [:person/id 7 :person/cars]     [[:car/id 5]]}

```

Reply:
The same logic can easily be repurposed for other helpers such as `remove-edge*`